File: 1594. maximum-non-negative-product-in-a-matrix.py
class Solution:
    def maxProductPath(self, grid):
        m, n = len(grid), len(grid[0])
        MOD = 10**9 + 7

        mx = [[0] * n for _ in range(m)]
        mn = [[0] * n for _ in range(m)]

        mx[0][0] = mn[0][0] = grid[0][0]

        # first row
        for j in range(1, n):
            mx[0][j] = mn[0][j] = mx[0][j - 1] * grid[0][j]

        # first column
        for i in range(1, m):
            mx[i][0] = mn[i][0] = mx[i - 1][0] * grid[i][0]

        for i in range(1, m):
            for j in range(1, n):
                x = grid[i][j]

                a = mx[i - 1][j] * x
                b = mn[i - 1][j] * x
                c = mx[i][j - 1] * x
                d = mn[i][j - 1] * x

                mx[i][j] = max(a, b, c, d)
                mn[i][j] = min(a, b, c, d)

        ans = mx[m - 1][n - 1]
        return -1 if ans < 0 else ans % MOD

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    grid1 = [[-1,-2,-3],[-2,-3,-3],[-3,-3,-2]]
    output1 = sol.maxProductPath(grid1)
    print(f"Example 1: grid={grid1}")
    print(f"Output: {output1}")
    print(f"Expected: -1")
    print()
    
    # Example 2
    grid2 = [[1,-2,1],[1,-2,1],[3,-4,1]]
    output2 = sol.maxProductPath(grid2)
    print(f"Example 2: grid={grid2}")
    print(f"Output: {output2}")
    print(f"Expected: 8")
    print()
    
    # Example 3
    grid3 = [[1,3],[0,-4]]
    output3 = sol.maxProductPath(grid3)
    print(f"Example 3: grid={grid3}")
    print(f"Output: {output3}")
    print(f"Expected: 0")

File: 1594. maximum-non-negative-product-in-a-matrix.java
class Solution {
    public int maxProductPath(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        long MOD = 1000000007;

        long[][] mx = new long[m][n];
        long[][] mn = new long[m][n];

        mx[0][0] = mn[0][0] = grid[0][0];

        // first row
        for (int j = 1; j < n; j++) {
            mx[0][j] = mn[0][j] = mx[0][j - 1] * grid[0][j];
        }

        // first column
        for (int i = 1; i < m; i++) {
            mx[i][0] = mn[i][0] = mx[i - 1][0] * grid[i][0];
        }

        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                long x = grid[i][j];

                long a = mx[i - 1][j] * x;
                long b = mn[i - 1][j] * x;
                long c = mx[i][j - 1] * x;
                long d = mn[i][j - 1] * x;

                mx[i][j] = Math.max(Math.max(a, b), Math.max(c, d));
                mn[i][j] = Math.min(Math.min(a, b), Math.min(c, d));
            }
        }

        long ans = mx[m - 1][n - 1];
        if (ans < 0) return -1;
        return (int)(ans % MOD);
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        int[][] grid1 = {{-1,-2,-3},{-2,-3,-3},{-3,-3,-2}};
        int output1 = sol.maxProductPath(grid1);
        System.out.println("Example 1: grid=[[-1,-2,-3],[-2,-3,-3],[-3,-3,-2]]");
        System.out.println("Output: " + output1);
        System.out.println("Expected: -1");
        System.out.println();
        
        // Example 2
        int[][] grid2 = {{1,-2,1},{1,-2,1},{3,-4,1}};
        int output2 = sol.maxProductPath(grid2);
        System.out.println("Example 2: grid=[[1,-2,1],[1,-2,1],[3,-4,1]]");
        System.out.println("Output: " + output2);
        System.out.println("Expected: 8");
        System.out.println();
        
        // Example 3
        int[][] grid3 = {{1,3},{0,-4}};
        int output3 = sol.maxProductPath(grid3);
        System.out.println("Example 3: grid=[[1,3],[0,-4]]");
        System.out.println("Output: " + output3);
        System.out.println("Expected: 0");
    }
}